controller/src/main/java/io/pravega/controller/task/Stream/StreamMetadataTasks.java

```diff
@@ -381,9 +402,25 @@ public void initializeStreamWriters(final EventStreamClientFactory clientFactory
                 });
     }
 
-    private CompletableFuture<Boolean> isTruncated(String scope, String stream, Map<Long, Long> streamCut, OperationContext context) {
-        return streamMetadataStore.getTruncationRecord(scope, stream, context, executor)
-                .thenApply(truncationProp -> !truncationProp.getObject().isUpdating() || !truncationProp.getObject().getStreamCut().equals(streamCut));
+    @VisibleForTesting
+    CompletableFuture<Boolean> isTruncated(String scope, String stream, Map<Long, Long> streamCut, OperationContext context) {
+        CompletableFuture<State> stateFuture = streamMetadataStore.getState(scope, stream, true, context, executor);
+        CompletableFuture<StreamTruncationRecord> configPropertyFuture
+                = streamMetadataStore.getTruncationRecord(scope, stream, context, executor).thenApply(VersionedMetadata::getObject);
+        return CompletableFuture.allOf(stateFuture, configPropertyFuture)
+                                .thenApply(v -> {
+                                    State state = stateFuture.join();
+                                    StreamTruncationRecord truncationRecord = configPropertyFuture.join();
+
+                                    // if property is updating and doesnt match our request, its a subsequent update
+                                    if (truncationRecord.isUpdating()) {
+                                        return !truncationRecord.getStreamCut().equals(streamCut);
+                                    } else {
+                                        // if truncate-barrier is not updating, then truncate is complete if property matches our expectation 
+                                        // and state is not updating 
+                                        return !(truncationRecord.getStreamCut().equals(streamCut) && state.equals(State.TRUNCATING));
```

Comment:
Something is not write here. The PR description says the following for `isTruncated`:

```
truncationProp.streamCut == truncationCut && !truncationProp.isUpdating
 && state!= UPDATING
```

while the predicate here is:

```
!(truncationRecord.getStreamCut().equals(streamCut) && state.equals(State.TRUNCATING))
```

In the case the stream cuts match and the state is `TRUNCATING`, then this returns false (not truncated), which seems right. But, if either the stream cuts do not match or the state isn't `TRUNCATING`, then it returns true (is truncated), which I'm not sure is correct. Is it correct to evaluate to true in the case the stream cuts do not match?

Reply:
When we submit a new truncation, we set the truncationRecord to updating with new streamcut in it. 

So first condition to check is `truncationProp.isUpdating` 

case 1 : truncationProp.isUpdating = true 
that means there is an ongoing truncation request. 
If ongoing request doesnt have the same streamCut as that we requested, then our request has definitely completed else our request is definitely incomplete. 

case 2: truncationProp.isUpdating = false
Otherwise ongoing truncation is complete. Stream State may or may not have been reset.
So, if `truncationRecord.streamCut` matches our request and state is truncating, then we completed the truncation but are yet to reset the state. we will return InProgress. 



If there is no ongoing truncationRequest

Reply:
The PR description was over simplification and incorrect. Corrected the PR description with exact checks

Reply:
I have pretty much the same problem with `IsTruncated`. Focusing on step 2, we have the following:

```
 truncationProp.streamCut == ourStreamCut && state!= TRUNCATING // if truncation record is updated but state is not reset ==> InProgress
    else ==> done
```

while in the code I see:

```
!(truncationRecord.getStreamCut().equals(streamCut) && state.equals(State.TRUNCATING))
```

In more abstract terms, I see in the description:

```
boolean isUpdated = A && !B
```

while the code says:

```
boolean isUpdated = !(A && B)
```

where:
`A`: `truncationRecord.getStreamCut().equals(streamCut)`
`B`: `state.equals(State.TRUNCATING)`

Reply:
let me first simplify and state in words when a work is done.. then we can tally it with code. 
Done conditions:
1. property.record != ourRecord   // this can only happen if our property update completed and at least one more subsequent update was performed after ours which may or may not have completed yet. 
So our update attempt is done. 
2. if property.record == ourRecord // this means no subsequent update was even attempted. now our update may or may not have completed. 
2.1 If property.isUpdating == true then our update of metadata is not complete. 
2.2 if state == truncating then our metadata update is complete but our state is not reset yet. 

Now lets tally it with the code for `update` and see if above is satisfied or not:
```
if (configProperty.isUpdating()) {
// this tests condition 2.1
    return !configProperty.getStreamConfiguration().equals(newConfig);
} else {
// this checks condition 2.2 and 1
    return !(configProperty.getStreamConfiguration().equals(newConfig) && state.equals(State.UPDATING)); 
```